File: partiql-plan/src/main/java/org/partiql/plan/rex/RexPathIndex.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;

import java.util.List;

/**
 * Logical path by index expression abstract base class.
 */
public abstract class RexPathIndex extends RexBase {

    /**
     * Creates a new RexPathIndex instance.
     * @param operand operand rex (operand 0)
     * @param index index rex
     * @return new RexPathIndex instance
     */
    @NotNull
    public static RexPathIndex create(@NotNull Rex operand, @NotNull Rex index) {
        return new Impl(operand, index);
    }

    /**
     * Gets the operand.
     * @return operand rex (operand 0)
     */
    @NotNull
    public abstract Rex getOperand();

    /**
     * Gets the index.
     * @return index rex
     */
    public abstract Rex getIndex();

    @Override
    @NotNull
    protected final RexType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getOperand());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitPathIndex(this, ctx);
    }

    private static class Impl extends RexPathIndex {

        private final Rex operand;
        private final Rex index;

        private Impl(@NotNull Rex operand, @NotNull Rex index) {
            this.operand = operand;
            this.index = index;
        }

        @NotNull
        @Override
        public Rex getOperand() {
            return operand;
        }

        @NotNull
        @Override
        public Rex getIndex() {
            return index;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexPathIndex.kt
package org.partiql.plan.rex

/**
 * Logical path index operator.
 */
public interface RexPathIndex : Rex {

    public fun getOperand(): Rex

    public fun getIndex(): Rex

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitPathIndex(this, ctx)
}

/**
 * Standard internal implementation for [RexPathIndex].
 */
internal class RexPathIndexImpl(operand: Rex, index: Rex, type: RexType) : RexPathIndex {

    // DO NOT USE FINAL
    private var _operand = operand
    private var _index = index
    private var _type = type

    override fun getOperand() = _operand

    override fun getIndex() = _index

    override fun getType(): RexType = _type

    override fun getChildren(): Collection<Rex> = listOf(_operand, _index)
}
